File: Sources/TypeCarrierCore/TextEditHistory.swift
public struct TextEditHistory: Sendable {
    private var undoStack: [String]
    private var redoStack: [String]
    private let limit: Int

    public init(limit: Int = 100) {
        self.undoStack = []
        self.redoStack = []
        self.limit = max(1, limit)
    }

    public var canUndo: Bool {
        !undoStack.isEmpty
    }

    public var canRedo: Bool {
        !redoStack.isEmpty
    }

    public mutating func recordChange(from oldValue: String, to newValue: String) {
        guard oldValue != newValue else {
            return
        }

        undoStack.append(oldValue)
        if undoStack.count > limit {
            undoStack.removeFirst(undoStack.count - limit)
        }
        redoStack.removeAll()
    }

    public mutating func undo(current: String) -> String? {
        guard let previous = undoStack.popLast() else {
            return nil
        }

        redoStack.append(current)
        return previous
    }

    public mutating func redo(current: String) -> String? {
        guard let next = redoStack.popLast() else {
            return nil
        }

        undoStack.append(current)
        return next
    }

    public mutating func reset() {
        undoStack.removeAll()
        redoStack.removeAll()
    }
}

File: Apps/Android/app/src/main/java/org/typecarrier/android/domain/TextEditHistory.kt
package org.typecarrier.android.domain

class TextEditHistory(limit: Int = 100) {
    private val undoStack = ArrayDeque<String>()
    private val redoStack = ArrayDeque<String>()
    private val limit = limit.coerceAtLeast(1)

    val canUndo: Boolean
        get() = undoStack.isNotEmpty()

    val canRedo: Boolean
        get() = redoStack.isNotEmpty()

    fun recordChange(from: String, to: String) {
        if (from == to) {
            return
        }

        undoStack.addLast(from)
        while (undoStack.size > limit) {
            undoStack.removeFirst()
        }
        redoStack.clear()
    }

    fun undo(current: String): String? {
        val previous = undoStack.removeLastOrNull() ?: return null
        redoStack.addLast(current)
        return previous
    }

    fun redo(current: String): String? {
        val next = redoStack.removeLastOrNull() ?: return null
        undoStack.addLast(current)
        return next
    }

    fun reset() {
        undoStack.clear()
        redoStack.clear()
    }
}
